### ISSUE TITLE
2.0 : Incorrect set of results returned if string value in where expression evaluates to true 

### ISSUE THREAD

**Issue by rajagp** [couchbase/couchbase-lite-core#392]
Originally in  ticket : https://github.com/couchbase/couchbase-lite-core/issues/330#issuecomment-363216387

Case 1: 
```
.where(Expression.property("email").and(Function.contains(Expression.property("email"), substring: Expression.string (".uk"))))
```
Expected : 
No matching rows. Al though the above expression evaluated to true

Observed: 
Giving me the following  results.
```
[["id": "landmark_16181", "name":"2 Willow Road", "email": "2willowroad@nationaltrust.org.uk", ["id":"landmark_16254", "name": "575 Wandsworth Road", "email": "575wandsworthroad@nationaltrust.org.uk"]]
```
Why is that the case ?

Case 2: 
Removing the dangling "email" property expression gives me the expected set ..
```
[["id": "hotel_10142", "name": "Glencoe Youth Hostel", "email": "glencoe@syha.org.uk"],[ "id": "hotel_10290", "name": "Hill House Holiday Cottage", "email": "owner@hillhousellanrhidian.co.uk"], ["id": "hotel_10291", "name": "Number 38 The Gower", "email": "julia@number38thegower.co.uk"], ["id": "hotel_11219", "name": "Sandy Patch", "email": "stay@holiday-harlech.co.uk"], ["id": "hotel_11315", "name": "The Granary at Roch Mill", "email": "granary@rochmill.co.uk"]])
```

**Comment by snej** [comment]
This is correct behavior according to SQLite, as I found while diagnosing the issue this split off from.

**tl;dr:** Interpreting strings as booleans is weird. Most strings are false, but strings with nonzero numeric prefixes -- in this case `"2willowroad@nationaltrust.org.uk"` -- are true.

Evaluating `Expression.property("email").and(...)` involves interpreting a string as a boolean. SQLite has no boolean type, just numbers, so it tries to parse the string as a number. Its number parser just stops when it reaches an illegal numeric character, returning whatever came before it, or zero if there were no digits. So `"2willow"` evaluates to 2, which is true.

This is counter-intuitive to be sure. But to change the way this works would require writing our own C function to do the conversion, and making our SQL query generator wrap it around every single expression that gets interpreted as a boolean. This would probably slow down queries; I don't know by how much.

**Comment by djpongh** [comment]
Filed: https://issues.couchbase.com/browse/DOC-3218 to clarify the documentation on this behavior.
